### PR TITLE
SqlMetadataPrivFlags remove Enum.HasFlag in favor of bitwise operations

### PR DESCRIPTION
**Pull request by cmeyertons**
Noticed this in the profiler -- IsNullable was allocating the enum struct because of Enum.HasFlag

There's pretty wide spread usage of Enum.HasFlag throughout the repo, but this is the only one occurring in the SqlBulkCopy critical path.

### CODE REVIEW

**Comment by cheenamalhotra** [comment]
Please also update the same for other flags while you're at it.
Search for "IsDifferentName" in the same class, there're 6 more of these.

**Comment by Wraith2** [comment]
Netcore or netfx? because in netcore I was specifically advised to use Enum.HasFlag because it's intrinsic and the compiler should recognise it and emit good assembly.

**Comment by roji** [comment]
Yeah, Enum.HasFlag is indeed recommended in .NET Core. Anyone using .NET Framework at this point can't really care that much about high-perf, given the huge gaps between Framework and Core.

**Comment by cmeyertons** [comment]
@roji @Wraith2  I'm seeing the allocations from SqlMetadataPriv.GetNullable (Enum.GetFlag) occur in the JetBrains DotTrace TimeLine Profiler while running `netcoreapp3.1-Release`.  It's normally pretty spot on about tracking memory allocations 

I did a quick benchmark and you guys are correct - it does not seem to be allocating in my corporate .NET 5 application:

```ini
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.1500 (1909/November2018Update/19H2)
Intel Xeon E-2176M CPU 2.70GHz, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.202
  [Host] : .NET Core 5.0 (CoreCLR 5.0.521.16609, CoreFX 5.0.521.16609), X64 RyuJIT

Job=InProcess  Toolchain=InProcessEmitToolchain  
```
|  Method | Effect |      Mean |     Error |    StdDev |    Median | Rank | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------- |------- |----------:|----------:|----------:|----------:|-----:|------:|------:|------:|----------:|
| Control |  First | 0.0002 ns | 0.0010 ns | 0.0009 ns | 0.0000 ns |    1 |     - |     - |     - |         - |
| Bitwise |  First | 0.0145 ns | 0.0173 ns | 0.0144 ns | 0.0127 ns |    1 |     - |     - |     - |         - |

I guess i'm still confused why it's showing up in `netcoreapp3.1-Release` DotTrace profiler

**Comment by Wraith2** [comment]
I've thought this thought a bit. The PR's when I was changing this are when I was working on System.Data.SqlClient in the corefx repository and at that time the only target was netcore and it made sense to use HasFlag because it's simpler and had been fixed for netcore not to have bad codegen. Working on Microsoft.Data.SqlClient is different because we've got the netfx codebase to think about and one of the long term goals here is to unify as much code as possible between the two so that netfx perf improves wherever possible alongside the direct target of netcore.

Codegen on netcore between the two is [similar enough](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGueBtAGIAbbAHNcAXT7d6DGADsArvgYAVGLgwBRZapABLBRhkdTPBgDkICmAwC8DAAxpzPEePIOG5V9QsWHmKk3qR+Ae6iYuRO3uQAPPExpgC+pjSmAA5QBgBu2Bh2Glq6KgwA+gBmUbgA3On+PA0WcsAQEMIMABJ4QQCyeADWDgB8DAAUVTUMAGTqmjp6AHRB5ACU9vbFiyorUeT1jdymre2dPbhBLNjCnfZjU+K4SxdB49ul+Huea4cpQA==) that it's unlikely to be the cause of a major performance disruption to use masking on netcore.

So HasFlag was the correct choice when it was made but direct masking is now the correct choice and I think that should be the overall policy in this codebase. I'll try to incorporate that into my future PR's as well.

**Review by Wraith2 — APPROVED** [review]
LGTM

**Comment by cheenamalhotra** [comment]
> Please also update the same for other flags while you're at it.
> Search for "IsDifferentName" in the same class, there're 6 more of these.

@cmeyertons would you be able to address this small request soon?

**Comment by Wraith2** [comment]
I can pick up that change in a separate PR if that's easier.

**Comment by cmeyertons** [comment]
@cheenamalhotra @Wraith2 earliest I can get back to this is Monday :/

**Comment by cheenamalhotra** [comment]
FYI - I've updated the PR with additional changes to unblock from review.

**Review by cheenamalhotra — APPROVED** [review]

**Review by DavoudEshtehari — APPROVED** [review]
Could you update the `Updateability` to use the related enum member instead of `0x3`?

```
        public byte Updatability
        {
            get => (byte)(flags & _SqlMetadataFlags.IsUpdatableMask);
            set => flags = (_SqlMetadataFlags)((value & (byte)_SqlMetadataFlags.IsUpdatableMask) | ((int)flags & (int)~_SqlMetadataFlags.IsUpdatableMask));
        }
```

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs [hunk 1]
```diff
@@ -427,6 +427,11 @@ internal _SqlMetaData(int ordinal) : base()
             this.ordinal = ordinal;
         }
 
+        private bool HasFlag(_SqlMetadataFlags flag)
+        {
+            return (flags & flag) != 0;
```
[the remark this answers is not in the document]
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserHelperClasses.cs:432]: 1. It's not relevant to this PR.
2. Even if we were to, it doesn't seem to be named off the chart. All variables here are similar in their naming convention.

If we want to resolve any naming convention issues, might as well do it not just to 1 property but all others too but separately in another PR.

**Review by JRahnama — APPROVED** [review]

**Pull request merged**